File: src/rust/runtime.rs
//! Runtime: Connects feature engine to live data sources
//! Fetches variables, computes features, and exposes functions

use crate::feature_engine::*;
use crate::types::*;
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};

/// Variable source types
#[derive(Debug, Clone)]
pub enum VarSource {
    Telemetry,           // From TelemetrySnapshot
    Computed,            // Derived feature
    External(String),    // External API/file
    Constant(f64),       // Static value
    Cached { ttl_ms: u64 },
}

/// Runtime variable definition
#[derive(Debug, Clone)]
pub struct RuntimeVar {
    pub name: String,
    pub source: VarSource,
    pub expression: Option<String>,  // For computed vars
    pub default: f64,
    pub min: Option<f64>,
    pub max: Option<f64>,
}

/// Function registry entry
#[derive(Clone)]
pub struct RuntimeFunc {
    pub name: String,
    pub arity: usize,
    pub func: Arc<dyn Fn(&[f64]) -> f64 + Send + Sync>,
}

impl std::fmt::Debug for RuntimeFunc {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RuntimeFunc")
            .field("name", &self.name)
            .field("arity", &self.arity)
            .finish()
    }
}

/// Cached value with expiration
#[derive(Debug, Clone)]
struct CachedValue {
    value: f64,
    expires_at: Instant,
}

/// Runtime context for feature evaluation
pub struct Runtime {
    feature_engine: FeatureEngine,
    variables: HashMap<String, RuntimeVar>,
    functions: HashMap<String, RuntimeFunc>,
    cache: RwLock<HashMap<String, CachedValue>>,
    telemetry: RwLock<Option<TelemetrySnapshot>>,
    budgets: RwLock<Option<ResourceBudgets>>,
}

impl Runtime {
    pub fn new() -> Self {
        let mut rt = Self {
            feature_engine: FeatureEngine::new(),
            variables: HashMap::new(),
            functions: HashMap::new(),
            cache: RwLock::new(HashMap::new()),
            telemetry: RwLock::new(None),
            budgets: RwLock::new(None),
        };
        rt.register_builtin_functions();
        rt.register_builtin_variables();
        rt
    }

    /// Register built-in functions
    fn register_builtin_functions(&mut self) {
        // Clamping
        self.register_function("clamp", 3, Arc::new(|args| {
            args[0].clamp(args[1], args[2])
        }));

        // Linear interpolation
        self.register_function("lerp", 3, Arc::new(|args| {
            args[0] + (args[1] - args[0]) * args[2]
        }));

        // Smoothstep
        self.register_function("smoothstep", 3, Arc::new(|args| {
            let t = ((args[0] - args[1]) / (args[2] - args[1])).clamp(0.0, 1.0);
            t * t * (3.0 - 2.0 * t)
        }));

        // Min/Max
        self.register_function("min", 2, Arc::new(|args| args[0].min(args[1])));
        self.register_function("max", 2, Arc::new(|args| args[0].max(args[1])));

        // Absolute difference
        self.register_function("absdiff", 2, Arc::new(|args| (args[0] - args[1]).abs()));

        // Normalize to range
        self.register_function("norm", 3, Arc::new(|args| {
            (args[0] - args[1]) / (args[2] - args[1]).max(1e-10)
        }));

        // Sigmoid
        self.register_function("sigmoid", 1, Arc::new(|args| {
            1.0 / (1.0 + (-args[0]).exp())
        }));

        // ReLU
        self.register_function("relu", 1, Arc::new(|args| args[0].max(0.0)));

        // Exponential moving average weight
        self.register_function("ema_weight", 1, Arc::new(|args| {
            2.0 / (args[0] + 1.0)
        }));
    }

    /// Register built-in variables from telemetry
    fn register_builtin_variables(&mut self) {
        let telemetry_vars = [
            ("cpu_util", 0.0, Some(0.0), Some(1.0)),
            ("gpu_util", 0.0, Some(0.0), Some(1.0)),
            ("frametime_ms", 16.67, Some(0.0), Some(1000.0)),
            ("temp_cpu", 50.0, Some(0.0), Some(120.0)),
            ("temp_gpu", 50.0, Some(0.0), Some(120.0)),
        ];

        for (name, default, min, max) in telemetry_vars {
            self.variables.insert(name.to_string(), RuntimeVar {
                name: name.to_string(),
                source: VarSource::Telemetry,
                expression: None,
                default,
                min,
                max,
            });
        }

        // Constants
        self.variables.insert("pi".to_string(), RuntimeVar {
            name: "pi".to_string(),
            source: VarSource::Constant(std::f64::consts::PI),
            expression: None,
            default: std::f64::consts::PI,
            min: None,
            max: None,
        });

        self.variables.insert("e".to_string(), RuntimeVar {
            name: "e".to_string(),
            source: VarSource::Constant(std::f64::consts::E),
            expression: None,
            default: std::f64::consts::E,
            min: None,
            max: None,
        });
    }

    /// Register a custom function
    pub fn register_function(
        &mut self,
        name: &str,
        arity: usize,
        func: Arc<dyn Fn(&[f64]) -> f64 + Send + Sync>,
    ) {
        self.functions.insert(name.to_string(), RuntimeFunc {
            name: name.to_string(),
            arity,
            func,
        });
    }

    /// Register a computed variable
    pub fn register_computed_var(&mut self, name: &str, expression: &str, default: f64) {
        self.variables.insert(name.to_string(), RuntimeVar {
            name: name.to_string(),
            source: VarSource::Computed,
            expression: Some(expression.to_string()),
            default,
            min: None,
            max: None,
        });
    }

    /// Update telemetry snapshot
    pub fn update_telemetry(&self, snapshot: TelemetrySnapshot) {
        let mut tel = self.telemetry.write().unwrap();
        *tel = Some(snapshot);
    }

    /// Update resource budgets
    pub fn update_budgets(&self, budgets: ResourceBudgets) {
        let mut b = self.budgets.write().unwrap();
        *b = Some(budgets);
    }

    /// Fetch a variable value
    pub fn fetch_var(&self, name: &str) -> Option<f64> {
        let var_def = self.variables.get(name)?;

        // Check cache first
        if let VarSource::Cached { ttl_ms } = &var_def.source {
            let cache = self.cache.read().unwrap();
            if let Some(cached) = cache.get(name) {
                if cached.expires_at > Instant::now() {
                    return Some(cached.value);
                }
            }
        }

        let value = match &var_def.source {
            VarSource::Constant(v) => *v,
            VarSource::Telemetry => self.fetch_telemetry_var(name)?,
            VarSource::Computed => self.compute_var(var_def)?,
            VarSource::External(_path) => var_def.default, // TODO: implement
            VarSource::Cached { .. } => self.compute_var(var_def)?,
        };

        // Apply bounds
        let bounded = match (var_def.min, var_def.max) {
            (Some(min), Some(max)) => value.clamp(min, max),
            (Some(min), None) => value.max(min),
            (None, Some(max)) => value.min(max),
            (None, None) => value,
        };

        Some(bounded)
    }

    /// Fetch multiple variables
    pub fn fetch_vars(&self, names: &[&str]) -> HashMap<String, f64> {
        names.iter()
            .filter_map(|&name| self.fetch_var(name).map(|v| (name.to_string(), v)))
            .collect()
    }

    /// Fetch all registered variables
    pub fn fetch_all_vars(&self) -> HashMap<String, f64> {
        self.variables.keys()
            .filter_map(|name| self.fetch_var(name).map(|v| (name.clone(), v)))
            .collect()
    }

    fn fetch_telemetry_var(&self, name: &str) -> Option<f64> {
        let tel = self.telemetry.read().unwrap();
        let snapshot = tel.as_ref()?;

        Some(match name {
            "cpu_util" => snapshot.cpu_util,
            "gpu_util" => snapshot.gpu_util,
            "frametime_ms" => snapshot.frametime_ms,
            "temp_cpu" => snapshot.temp_cpu as f64,
            "temp_gpu" => snapshot.temp_gpu as f64,
            _ => return None,
        })
    }

    fn compute_var(&self, var: &RuntimeVar) -> Option<f64> {
        let expr_str = var.expression.as_ref()?;

        // Load current variables into engine
        let vars = self.fetch_base_vars();
        let mut engine = FeatureEngine::new();
        for (k, v) in vars {
            engine.set_variable(&k, v);
        }

        // Parse and evaluate
        let expr = engine.parse(expr_str).ok()?;
        engine.evaluate(&expr).ok()
    }

    fn fetch_base_vars(&self) -> HashMap<String, f64> {
        let mut vars = HashMap::new();

        if let Some(snapshot) = self.telemetry.read().unwrap().as_ref() {
            vars.insert("cpu_util".to_string(), snapshot.cpu_util);
            vars.insert("gpu_util".to_string(), snapshot.gpu_util);
            vars.insert("frametime_ms".to_string(), snapshot.frametime_ms);
            vars.insert("temp_cpu".to_string(), snapshot.temp_cpu as f64);
            vars.insert("temp_gpu".to_string(), snapshot.temp_gpu as f64);
        }

        vars.insert("pi".to_string(), std::f64::consts::PI);
        vars.insert("e".to_string(), std::f64::consts::E);

        vars
    }

    /// Call a registered function
    pub fn call_function(&self, name: &str, args: &[f64]) -> Option<f64> {
        let func = self.functions.get(name)?;
        if args.len() != func.arity {
            return None;
        }
        Some((func.func)(args))
    }

    /// Evaluate an expression with current runtime state
    pub fn evaluate(&self, expression: &str) -> Result<f64, String> {
        let vars = self.fetch_all_vars();
        let mut engine = FeatureEngine::new();
        for (k, v) in vars {
            engine.set_variable(&k, v);
        }
        let expr = engine.parse(expression)?;
        engine.evaluate(&expr)
    }

    /// Compute a feature with scaling
    pub fn compute_scaled_feature(
        &self,
        base_var: &str,
        params: &ScaleParams,
    ) -> Option<f64> {
        let value = self.fetch_var(base_var)?;
        Some(self.feature_engine.scale_abt(value, params))
    }

    /// Compute multiple features from definitions
    pub fn compute_features(&self, definitions: &[(String, String)]) -> HashMap<String, f64> {
        definitions.iter()
            .filter_map(|(name, expr)| {
                self.evaluate(expr).ok().map(|v| (name.clone(), v))
            })
            .collect()
    }

    /// List all registered variables
    pub fn list_variables(&self) -> Vec<&str> {
        self.variables.keys().map(|s| s.as_str()).collect()
    }

    /// List all registered functions
    pub fn list_functions(&self) -> Vec<(&str, usize)> {
        self.functions.iter().map(|(k, v)| (k.as_str(), v.arity)).collect()
    }
}

impl Default for Runtime {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_telemetry() -> TelemetrySnapshot {
        TelemetrySnapshot {
            timestamp: "2025-01-21T14:30:00Z".into(),
            cpu_util: 0.75,
            gpu_util: 0.70,
            frametime_ms: 14.2,
            temp_cpu: 72,
            temp_gpu: 68,
            active_process_category: "gaming".into(),
        }
    }

    #[test]
    fn test_fetch_telemetry_var() {
        let rt = Runtime::new();
        rt.update_telemetry(sample_telemetry());

        assert!((rt.fetch_var("cpu_util").unwrap() - 0.75).abs() < 0.001);
        assert!((rt.fetch_var("temp_cpu").unwrap() - 72.0).abs() < 0.001);
    }

    #[test]
    fn test_computed_var() {
        let mut rt = Runtime::new();
        rt.update_telemetry(sample_telemetry());
        rt.register_computed_var("cpu_scaled", "cpu_util * 2.0", 0.0);

        assert!((rt.fetch_var("cpu_scaled").unwrap() - 1.5).abs() < 0.001);
    }

    #[test]
    fn test_call_function() {
        let rt = Runtime::new();

        assert!((rt.call_function("clamp", &[1.5, 0.0, 1.0]).unwrap() - 1.0).abs() < 0.001);
        assert!((rt.call_function("lerp", &[0.0, 10.0, 0.5]).unwrap() - 5.0).abs() < 0.001);
        assert!((rt.call_function("sigmoid", &[0.0]).unwrap() - 0.5).abs() < 0.001);
    }

    #[test]
    fn test_evaluate_expression() {
        let rt = Runtime::new();
        rt.update_telemetry(sample_telemetry());

        let result = rt.evaluate("cpu_util + gpu_util").unwrap();
        assert!((result - 1.45).abs() < 0.001);
    }
}

File: src/python/runtime.py
"""
Runtime: Connects feature engine to live data sources.
Fetches variables, computes features, and exposes functions.
"""

import time
import math
from typing import Dict, List, Optional, Callable, Any, Tuple
from dataclasses import dataclass, field
from enum import Enum
from threading import RLock
from datetime import datetime

from .feature_engine import FeatureEngine, ScaleParams
from .schemas import TelemetrySnapshot


class VarSource(Enum):
    TELEMETRY = "telemetry"
    COMPUTED = "computed"
    EXTERNAL = "external"
    CONSTANT = "constant"
    CACHED = "cached"


@dataclass
class RuntimeVar:
    """Runtime variable definition."""
    name: str
    source: VarSource
    expression: Optional[str] = None
    default: float = 0.0
    min_val: Optional[float] = None
    max_val: Optional[float] = None
    ttl_ms: Optional[int] = None  # For cached vars


@dataclass
class CachedValue:
    """Cached value with expiration."""
    value: float
    expires_at: float  # Unix timestamp


@dataclass
class RuntimeFunc:
    """Registered function."""
    name: str
    arity: int
    func: Callable[..., float]


class Runtime:
    """
    Runtime context for feature evaluation.
    Connects feature engine to live telemetry and provides function registry.
    """

    def __init__(self):
        self.feature_engine = FeatureEngine()
        self.variables: Dict[str, RuntimeVar] = {}
        self.functions: Dict[str, RuntimeFunc] = {}
        self.cache: Dict[str, CachedValue] = {}
        self.telemetry: Optional[TelemetrySnapshot] = None
        self.budgets: Optional[Dict[str, Any]] = None
        self._lock = RLock()

        self._register_builtin_functions()
        self._register_builtin_variables()

    def _register_builtin_functions(self) -> None:
        """Register built-in utility functions."""
        # Clamping
        self.register_function("clamp", 3, lambda x, lo, hi: max(lo, min(hi, x)))

        # Linear interpolation
        self.register_function("lerp", 3, lambda a, b, t: a + (b - a) * t)

        # Smoothstep
        def smoothstep(x, edge0, edge1):
            t = max(0, min(1, (x - edge0) / (edge1 - edge0)))
            return t * t * (3 - 2 * t)
        self.register_function("smoothstep", 3, smoothstep)

        # Min/Max
        self.register_function("min", 2, min)
        self.register_function("max", 2, max)

        # Absolute difference
        self.register_function("absdiff", 2, lambda a, b: abs(a - b))

        # Normalize to range
        self.register_function("norm", 3, lambda x, lo, hi: (x - lo) / max(hi - lo, 1e-10))

        # Sigmoid
        self.register_function("sigmoid", 1, lambda x: 1.0 / (1.0 + math.exp(-x)))

        # ReLU
        self.register_function("relu", 1, lambda x: max(0, x))

        # EMA weight
        self.register_function("ema_weight", 1, lambda n: 2.0 / (n + 1.0))

        # Thermal margin
        self.register_function("thermal_margin", 2, lambda temp, max_temp: max(0, max_temp - temp))

        # FPS from frametime
        self.register_function("fps", 1, lambda ft: 1000.0 / max(ft, 0.001))

    def _register_builtin_variables(self) -> None:
        """Register built-in variables from telemetry."""
        telemetry_vars = [
            ("cpu_util", 0.0, 0.0, 1.0),
            ("gpu_util", 0.0, 0.0, 1.0),
            ("frametime_ms", 16.67, 0.0, 1000.0),
            ("temp_cpu", 50.0, 0.0, 120.0),
            ("temp_gpu", 50.0, 0.0, 120.0),
        ]

        for name, default, min_v, max_v in telemetry_vars:
            self.variables[name] = RuntimeVar(
                name=name,
                source=VarSource.TELEMETRY,
                default=default,
                min_val=min_v,
                max_val=max_v,
            )

        # Constants
        self.variables["pi"] = RuntimeVar(
            name="pi", source=VarSource.CONSTANT, default=math.pi
        )
        self.variables["e"] = RuntimeVar(
            name="e", source=VarSource.CONSTANT, default=math.e
        )

    def register_function(self, name: str, arity: int, func: Callable[..., float]) -> None:
        """Register a custom function."""
        self.functions[name] = RuntimeFunc(name=name, arity=arity, func=func)

    def register_computed_var(
        self,
        name: str,
        expression: str,
        default: float = 0.0,
        min_val: Optional[float] = None,
        max_val: Optional[float] = None,
    ) -> None:
        """Register a computed variable with an expression."""
        self.variables[name] = RuntimeVar(
            name=name,
            source=VarSource.COMPUTED,
            expression=expression,
            default=default,
            min_val=min_val,
            max_val=max_val,
        )

    def register_cached_var(
        self,
        name: str,
        expression: str,
        ttl_ms: int = 1000,
        default: float = 0.0,
    ) -> None:
        """Register a cached computed variable."""
        self.variables[name] = RuntimeVar(
            name=name,
            source=VarSource.CACHED,
            expression=expression,
            default=default,
            ttl_ms=ttl_ms,
        )

    def update_telemetry(self, snapshot: TelemetrySnapshot) -> None:
        """Update telemetry snapshot."""
        with self._lock:
            self.telemetry = snapshot

    def update_telemetry_dict(self, data: Dict[str, Any]) -> None:
        """Update telemetry from a dictionary."""
        with self._lock:
            self.telemetry = TelemetrySnapshot(
                timestamp=datetime.fromisoformat(data.get("timestamp", datetime.now().isoformat())),
                cpu_util=data.get("cpu_util", 0.0),
                gpu_util=data.get("gpu_util", 0.0),
                frametime_ms=data.get("frametime_ms", 16.67),
                temp_cpu=data.get("temp_cpu", 50),
                temp_gpu=data.get("temp_gpu", 50),
                active_process_category=data.get("active_process_category", "unknown"),
            )

    def fetch_var(self, name: str) -> Optional[float]:
        """Fetch a variable value."""
        with self._lock:
            var_def = self.variables.get(name)
            if not var_def:
                return None

            # Check cache for cached vars
            if var_def.source == VarSource.CACHED:
                cached = self.cache.get(name)
                if cached and cached.expires_at > time.time():
                    return cached.value

            # Fetch based on source
            if var_def.source == VarSource.CONSTANT:
                value = var_def.default
            elif var_def.source == VarSource.TELEMETRY:
                value = self._fetch_telemetry_var(name)
            elif var_def.source in (VarSource.COMPUTED, VarSource.CACHED):
                value = self._compute_var(var_def)
            else:
                value = var_def.default

            if value is None:
                value = var_def.default

            # Apply bounds
            if var_def.min_val is not None:
                value = max(var_def.min_val, value)
            if var_def.max_val is not None:
                value = min(var_def.max_val, value)

            # Update cache for cached vars
            if var_def.source == VarSource.CACHED and var_def.ttl_ms:
                self.cache[name] = CachedValue(
                    value=value,
                    expires_at=time.time() + var_def.ttl_ms / 1000.0,
                )

            return value

    def fetch_vars(self, names: List[str]) -> Dict[str, float]:
        """Fetch multiple variables."""
        return {name: v for name in names if (v := self.fetch_var(name)) is not None}

    def fetch_all_vars(self) -> Dict[str, float]:
        """Fetch all registered variables."""
        return self.fetch_vars(list(self.variables.keys()))

    def _fetch_telemetry_var(self, name: str) -> Optional[float]:
        """Fetch a variable from telemetry."""
        if not self.telemetry:
            return None

        mapping = {
            "cpu_util": self.telemetry.cpu_util,
            "gpu_util": self.telemetry.gpu_util,
            "frametime_ms": self.telemetry.frametime_ms,
            "temp_cpu": float(self.telemetry.temp_cpu),
            "temp_gpu": float(self.telemetry.temp_gpu),
        }
        return mapping.get(name)

    def _compute_var(self, var: RuntimeVar) -> Optional[float]:
        """Compute a variable from its expression."""
        if not var.expression:
            return var.default

        try:
            return self.evaluate(var.expression)
        except Exception:
            return var.default

    def call_function(self, name: str, *args: float) -> Optional[float]:
        """Call a registered function."""
        func = self.functions.get(name)
        if not func:
            return None
        if len(args) != func.arity:
            return None
        return func.func(*args)

    def evaluate(self, expression: str) -> float:
        """Evaluate an expression with current runtime state."""
        # Set all variables in engine
        self.feature_engine.set_variables(self._get_base_vars())
        return self.feature_engine.parse_and_evaluate(expression)

    def _get_base_vars(self) -> Dict[str, float]:
        """Get base variables for expression evaluation."""
        vars_dict: Dict[str, float] = {"pi": math.pi, "e": math.e}

        if self.telemetry:
            vars_dict.update({
                "cpu_util": self.telemetry.cpu_util,
                "gpu_util": self.telemetry.gpu_util,
                "frametime_ms": self.telemetry.frametime_ms,
                "temp_cpu": float(self.telemetry.temp_cpu),
                "temp_gpu": float(self.telemetry.temp_gpu),
            })

        return vars_dict

    def compute_scaled_feature(self, base_var: str, params: ScaleParams) -> Optional[float]:
        """Compute a feature with alpha-beta-theta scaling."""
        value = self.fetch_var(base_var)
        if value is None:
            return None
        return self.feature_engine.scale_abt(value, params)

    def compute_features(self, definitions: List[Tuple[str, str]]) -> Dict[str, float]:
        """Compute multiple features from definitions."""
        results = {}
        for name, expr in definitions:
            try:
                results[name] = self.evaluate(expr)
            except Exception:
                pass
        return results

    def list_variables(self) -> List[str]:
        """List all registered variable names."""
        return list(self.variables.keys())

    def list_functions(self) -> List[Tuple[str, int]]:
        """List all registered functions with arity."""
        return [(f.name, f.arity) for f in self.functions.values()]

    def get_runtime_state(self) -> Dict[str, Any]:
        """Get current runtime state for debugging/logging."""
        return {
            "variables": self.fetch_all_vars(),
            "telemetry_active": self.telemetry is not None,
            "cache_size": len(self.cache),
            "registered_vars": len(self.variables),
            "registered_funcs": len(self.functions),
        }


# Convenience function for quick evaluation
def quick_eval(expression: str, **variables: float) -> float:
    """Quickly evaluate an expression with given variables."""
    rt = Runtime()
    for name, value in variables.items():
        rt.variables[name] = RuntimeVar(
            name=name, source=VarSource.CONSTANT, default=value
        )
    return rt.evaluate(expression)


if __name__ == "__main__":
    # Demo
    rt = Runtime()

    # Simulate telemetry update
    rt.update_telemetry_dict({
        "cpu_util": 0.75,
        "gpu_util": 0.70,
        "frametime_ms": 14.2,
        "temp_cpu": 72,
        "temp_gpu": 68,
    })

    # Register computed features
    rt.register_computed_var("cpu_scaled", "cpu_util * 1.5 + 0.1")
    rt.register_computed_var("thermal_risk", "sigmoid((temp_cpu - 80) / 10)")
    rt.register_computed_var("fps", "1000 / frametime_ms")

    print("Runtime Demo")
    print("=" * 40)
    print(f"Variables: {rt.list_variables()}")
    print(f"Functions: {rt.list_functions()}")
    print()
    print("Fetched values:")
    for name in ["cpu_util", "gpu_util", "cpu_scaled", "thermal_risk", "fps"]:
        print(f"  {name}: {rt.fetch_var(name):.4f}")
    print()
    print("Expression evaluation:")
    print(f"  cpu_util + gpu_util = {rt.evaluate('cpu_util + gpu_util'):.4f}")
    print(f"  sin(cpu_util * pi) = {rt.evaluate('sin(cpu_util * pi)'):.4f}")
    print()
    print("Function calls:")
    print(f"  clamp(1.5, 0, 1) = {rt.call_function('clamp', 1.5, 0.0, 1.0)}")
    print(f"  sigmoid(0) = {rt.call_function('sigmoid', 0.0)}")
